stardist/bioimageio_utils.py

```diff
@@ -0,0 +1,372 @@
+    net_axes_out = axes_check_and_normalize(model._axes_out).lower()
+    ndim_tensor = len(net_axes_out) + 1
+
+    input_min_shape = list(axes_net_div_by)
+    input_min_shape[axes_net.index('C')] = model.config.n_channel_in
+    input_step = list(axes_net_div_by)
+    input_step[axes_net.index('C')] = 0
+
+    # the axes strings in bioimageio convention
+    input_axes = "b" + net_axes_in.lower()
+    output_axes = "b" + net_axes_out.lower()
+
+    if mode == "keras_hdf5":
+        output_names = ("prob", "dist") + (("class_prob",) if model._is_multiclass() else ())
+        output_n_channels = (1, model.config.n_rays,) + ((1,) if model._is_multiclass() else ())
+        # the output shape is computed from the input shape using
+        # output_shape[i] = output_scale[i] * input_shape[i] + 2 * output_offset[i]
+        output_scale = [1]+list(1/g for g in model.config.grid) + [0]
+        output_offset = [0]*(ndim_tensor)
+
+    elif mode == "tensorflow_saved_model_bundle":
+        if model._is_multiclass():
+            raise NotImplementedError("Tensorflow SavedModel not supported for multiclass models yet")
+        # regarding input/output names: https://github.com/CSBDeep/CSBDeep/blob/b0d2f5f344ebe65a9b4c3007f4567fe74268c813/csbdeep/utils/tf.py#L193-L194
+        input_names  = ["input"]
+        output_names = ["output"]
+        output_n_channels = (1 + model.config.n_rays,)
+        # the output shape is computed from the input shape using
+        # output_shape[i] = output_scale[i] * input_shape[i] + 2 * output_offset[i]
+        # same shape as input except for the channel dimension
+        output_scale = [1]*(ndim_tensor)
+        output_scale[output_axes.index("c")] = 0
+        # no offset, except for the input axes, where it is output channel / 2
+        output_offset = [0.0]*(ndim_tensor)
+        output_offset[output_axes.index("c")] = output_n_channels[0] / 2.0
+
+    # TODO need config format that is compatible with deepimagej; discuss with Esti
+    # TODO do we need parameters for down/upsampling here?
+    metadata, *_ = _import()
+    package_data = metadata("stardist")
+    config = dict(
+        stardist=dict(
+            python_version=package_data["Version"],
+            thresholds=dict(nms=model.thresholds.nms, prob=model.thresholds.prob),
+            config=vars(model.config),
+        )
+    )
+
+    n_inputs = len(input_names)
+    assert n_inputs == 1
+    input_config = dict(
+        input_name=input_names,
+        input_min_shape=[[1]+input_min_shape],
+        input_step=[[0]+input_step],
+        input_axes=[input_axes],
+        input_data_range=[["-inf", "inf"]],
+        preprocessing=[dict(scale_range=dict(
+            mode="per_sample",
+            axes=axes_norm.lower(),
+            min_percentile=min_percentile,
+            max_percentile=max_percentile,
+        ))]
+    )
+
+    n_outputs = len(output_names)
+    halo = [0 if ax in "bc" else 32 for ax in output_axes]
```

Comment:
Why is the `halo` defined with respect to the output? How is this used to calculate the amount of overlap to be used for tiles of the input? (when doing tiled processing)

Reply:
In particular, does this calculation make use of `output_scale` etc.?

Reply:
> Why is the `halo` defined with respect to the output?

Because the halo is cropped from the output and not the input, so it makes more sense to specify it like this.
I.e. the full tile is predicted and then the halo is removed after prediction. This only makes a difference if we have an `output_scale`.

> In particular, does this calculation make use of `output_scale` etc.?

The output shape is computed from the input shape using `output_scale`. The `halo` is given in the output coordinates, so to determine the size of the halo cropped from the input you would need to apply the inverse `output_scale` to it. 

That being said, we haven't implemented tiling with an output_scale yet, see
https://github.com/bioimage-io/core-bioimage-io-python/blob/main/bioimageio/core/prediction.py#L443-L446

(But I plan on implementing this soon.)